pkg/jaeger/store/tag_info.go

```diff
@@ -0,0 +1,168 @@
+			tagsInfo.params = append(tagsInfo.params, "opencensus.exporterversion", "\"Jaeger-"+v+"\"")
+			qual := fmt.Sprintf(`s.span_tags @> (SELECT _ps_trace.tag_v_eq_matching_tags($%d::text, $%d::jsonb))`, len(tagsInfo.params)-1, len(tagsInfo.params))
+			tagsInfo.spanClauses = append(tagsInfo.spanClauses, qual)
+		case TagSpanKind:
+			tagsInfo.params = append(tagsInfo.params, jSpanKindToInternal(v))
+			qual := fmt.Sprintf("op.span_kind = $%d", len(tagsInfo.params))
+			tagsInfo.operationClauses = append(tagsInfo.operationClauses, qual)
+		case TagW3CTraceState:
+			tagsInfo.params = append(tagsInfo.params, v)
+			qual := fmt.Sprintf(`s.trace_state = $%d`, len(tagsInfo.params))
+			tagsInfo.spanClauses = append(tagsInfo.spanClauses, qual)
+		case TagEventName:
+			tagsInfo.params = append(tagsInfo.params, v)
+			qual := fmt.Sprintf(`e.name = $%d`, len(tagsInfo.params))
+			tagsInfo.eventClauses = append(tagsInfo.eventClauses, qual)
+		default:
+			tagsInfo.generalTags = append(tagsInfo.generalTags, &tag{k: k, v: v})
+		}
+	}
+
+	if len(tagsInfo.generalTags) > 0 {
+		query, params := tagsInfo.BuildTagsQuery(q)
+		rows, err := conn.Query(ctx, query, params...)
+		if err != nil {
+			return nil, fmt.Errorf("querying traces error: %w query:\n%s", err, query)
+		}
+		defer rows.Close()
+		possibleToMatch, err := tagsInfo.scanTags(rows)
+		if err != nil {
+			return nil, fmt.Errorf("scanning traces error: %w query:\n%s", err, query)
+		}
+		if !possibleToMatch {
+			return nil, nil
+		}
+	}
+
+	return tagsInfo, nil
+}
+
+// BuildTagsQuery returns the query to be used to scan tags.
+// Exposed for end to end tests.
+func (t *tagsInfo) BuildTagsQuery(q *spanstore.TraceQueryParameters) (string, []interface{}) {
+	params := make([]interface{}, 0, 2)
+
+	keys := make([]string, len(t.generalTags))
+	values := make([]string, len(t.generalTags))
+	for i, tag := range t.generalTags {
+		keys[i] = tag.k
+		values[i] = tag.v
+	}
+	params = append(params, keys, values)
+	c := fmt.Sprintf(`
+		SELECT result.*
+		FROM unnest ($%d::text[], $%d::text[]) with ordinality as tag_qual(key, value, nr)
+		INNER JOIN LATERAL
+		(
+			SELECT
+				array_agg(pg_catalog.jsonb_build_object(tag.key_id, tag.id)) json_pair_array,
+				(bit_or(tag.tag_type) OPERATOR(pg_catalog.&) ps_trace.span_tag_type() != 0) is_span_type,
+				(bit_or(tag.tag_type) OPERATOR(pg_catalog.&) ps_trace.resource_tag_type() != 0) is_resource_type,
+				(bit_or(tag.tag_type) OPERATOR(pg_catalog.&) ps_trace.event_tag_type() != 0) is_event_type
+			FROM
+			(
+					SELECT v, _prom_ext.jsonb_digest(v) as digest
+					FROM unnest(_ps_trace.text_matches(tag_qual.value)) as v
+			) as matchers
+			INNER JOIN LATERAL (
+				SELECT t.*
+				FROM _ps_trace.tag t
+				WHERE t.key OPERATOR(pg_catalog.=) tag_qual.key
+				AND _prom_ext.jsonb_digest(t.value) OPERATOR(pg_catalog.=) matchers.digest
+				AND t.value OPERATOR(pg_catalog.=) matchers.v
```

Comment:
Isn't value comparison and value digest comparison would be always true?

Reply:
This is to future-proof against possible collisions